### PR TITLE
fix(Core/Spell): Spell Lock

### PR DESCRIPTION
**Pull request by grimgravy**
According to @Kitzunu, the problem is in the order of execution of the effect.

### the order of effect
the problem: spell trigger (spell_effect_2) ignores mechanic: interrupted (spell_effect_1)

spell_effect_1

the solution to the problem is to leave spell_effect_1, in first order
in second order spell_effect_2 by the script.
this way, you solve this problem 

spell_effect_2
### test
[Youtube](https://www.youtube.com/watch?v=0LvMElwh4ts)
## Issues Addressed:

- Closes [#8484](https://github.com/azerothcore/azerothcore-wotlk/issues/8484)
- Closes [#2026](https://github.com/chromiecraft/chromiecraft/issues/2026)

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by grimgravy** [comment]
I will change later

**Comment by FrancescoBorzi** [comment]
@grimgravy any news about this? still WIP?

**Comment by Nyeriah** [comment]
Just an update on this: I've tried https://github.com/TrinityCore/TrinityCore/commit/ed6cf717c8ffafe9bf14298d3582c171008c8b49 and sadly it didn't work, we might still be missing more spell system changes, maybe procs.

**Comment by grimgravy** [comment]
Nyeriah@, I did the same procedure as you, and it worked. However, the "EffectInterruptCast" change is not the correct way

**Comment by grimgravy** [comment]
> @grimgravy any news about this? still WIP?

lack of test

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers can anyone please check this?

**Comment by claudiodfc** [comment]
Tested.
- Warlock at level 36, which has the rank 1 of `Spell Lock`, vs a Mage casting Frostfire Bolt silences the mage for the duration of the `Spell Lock` effect.

File: src/server/game/Spells/SpellEffects.cpp [hunk 1]
```diff
@@ -799,15 +799,15 @@ void Spell::EffectDummy(SpellEffIndex effIndex)
 
 void Spell::EffectTriggerSpell(SpellEffIndex effIndex)
 {
-    if (effectHandleMode != SPELL_EFFECT_HANDLE_LAUNCH_TARGET
+    if (effectHandleMode != SPELL_EFFECT_HANDLE_HIT_TARGET
```
Comment (UltraNix) [inline src/server/game/Spells/SpellEffects.cpp:802]: I don't like that change - might break many spells using that effect. `SPELL_EFFECT_HANDLE_HIT_TARGET` might be `too late` hook for triggered spells.

**Review by UltraNix — CHANGES_REQUESTED** [review]

File: src/server/game/Spells/SpellEffects.cpp (hunk 1 above)
Comment (FrancescoBorzi) [reply src/server/game/Spells/SpellEffects.cpp:802]: @grimgravy

Comment (grimgravy) [reply src/server/game/Spells/SpellEffects.cpp:802]: I didn't see a problem related to this

Comment (UltraNix) [reply src/server/game/Spells/SpellEffects.cpp:802]: There is a difference between launch and hit. And troggered spells must be launched with the base spell. In other case, all delayed spells would be broken.

**Comment by FrancescoBorzi** [comment]
@grimgravy what's the status of this?

File: src/server/game/Spells/SpellEffects.cpp (hunk 1 above)
Comment (grimgravy) [reply src/server/game/Spells/SpellEffects.cpp:802]: I will make a change soon.

Comment (grimgravy) [reply src/server/game/Spells/SpellEffects.cpp:802]: UltraNix@, this is very logical, isn't it? If it continues like this there is no solution.

Comment (UltraNix) [reply src/server/game/Spells/SpellEffects.cpp:802]: What is logical?

Comment (grimgravy) [reply src/server/game/Spells/SpellEffects.cpp:802]: @UltraNix, what is difference for you?

Comment (UltraNix) [reply src/server/game/Spells/SpellEffects.cpp:802]: `There is a difference between launch and hit. And triggered spells must be launched with the base spell. In other case, all delayed spells would be broken.`

**Comment by claudiodfc** [comment]
As far as I can see, the problem lies in the change itself. Currently the core is preventing spells that aren't in `SPELL_EFFECT_HANDLE_LAUNCH_TARGET` to be triggered and the commit is changing that to `SPELL_EFFECT_HANDLE_HIT_TARGET`. If you think about it, there's spells that are triggered at launched and not on hit. You could try to change it to the bellow snippet and test other triggered spells.
```
(SPELL_EFFECT_HANDLE_HIT_TARGET || SPELL_EFFECT_HANDLE_LAUNCH_TARGET)
```

**Comment by FrancescoBorzi** [comment]
closing as no answer from the author, it can be reopened if someone wants to keep working on this

**Pull request closed without merging**